### DISCUSSION TITLE
Proper way to test this ?

### DISCUSSION THREAD

**Discussion by drupol** [infection/infection#1689] [Q&A] [answered]
Hello,

I'm trying to increase my coverage in my project. I noticed that there is a mutant escaping despite the fact that the test should cover it.

![image](https://user-images.githubusercontent.com/252042/167094122-dda915f2-e6c3-4342-9a4b-4a3d3817308d.png)

Link: https://dashboard.stryker-mutator.io/reports/github.com/loophp/collection/master#mutant/Utils/CallbacksArrayReducer.php

The [class](https://github.com/loophp/collection/blob/master/src/Utils/CallbacksArrayReducer.php) is:

<details>

```php
<?php

declare(strict_types=1);

namespace loophp\collection\Utils;

use Closure;

/**
 * @internal
 *
 * @immutable
 *
 * phpcs:disable Generic.Files.LineLength.TooLong
 */
final class CallbacksArrayReducer
{
    /**
     * @return Closure(array<array-key, callable(mixed...): bool>): Closure(mixed...): bool
     */
    public static function or(): Closure
    {
        return
            /**
             * @param array<array-key, callable(mixed...): bool> $callbacks
             *
             * @return Closure(mixed...): bool
             */
            static fn (array $callbacks): Closure =>
                /**
                 * @param mixed ...$parameters
                 */
                static fn (...$parameters): bool => array_reduce(
                    $callbacks,
                    /**
                     * @param bool $carry
                     * @param callable(mixed...): bool $callable
                     */
                    static fn (bool $carry, callable $callable): bool => $carry || $callable(...$parameters),
                    false
                );
    }
}
```

</details>

The [PHPUnit tests](https://github.com/loophp/collection/blob/master/tests/unit/Utils/CallbacksArrayReducerTest.php) are:

<details>

```php
<?php

declare(strict_types=1);

namespace tests\loophp\collection\Utils;

use loophp\collection\Utils\CallbacksArrayReducer;
use PHPUnit\Framework\TestCase;

/**
 * @internal
 * @coversDefaultClass \loophp\collection\Utils\CallbacksArrayReducer
 */
final class CallbacksArrayReducerTest extends TestCase
{
    public function dataProvider()
    {
        // Ensure Callbacks receive the needed arguments.
        yield [
            [
                static fn (string $value, string $key, iterable $iterable): bool => 'value_key_a_b_c' === sprintf('%s_%s_%s', $value, $key, implode('_', $iterable)),
            ],
            'value',
            'key',
            range('a', 'c'),
            true,
        ];

        // Reduce empty callbacks array
        yield [
            [],
            0,
            0,
            range('a', 'c'),
            false,
        ];

        // Reduce multiple callbacks
        yield [
            [
                static fn (int $v): bool => 5 < $v,
                static fn (int $v): bool => 0 === $v % 2,
            ],
            0,
            0,
            range(0, 10),
            true,
        ];

        yield [
            [
                static fn (int $v): bool => 5 < $v,
                static fn (int $v): bool => 0 === $v % 2,
            ],
            13,
            0,
            range(0, 10),
            true,
        ];

        yield [
            [
                static fn (int $v): bool => 5 < $v,
                static fn (int $v): bool => 0 === $v % 2,
            ],
            3,
            0,
            range(0, 10),
            false,
        ];

        // Reduce single callback
        yield [
            [
                static fn (int $v): bool => 5 < $v,
            ],
            0,
            0,
            range(0, 10),
            false,
        ];

        yield [
            [
                static fn (int $v): bool => 5 < $v,
            ],
            6,
            0,
            range(0, 10),
            true,
        ];

        yield [
            [
                static fn (): bool => false,
            ],
            [],
            [],
            [],
            false,
        ];
    }

    /**
     * @dataProvider dataProvider
     *
     * @param mixed $current
     * @param mixed $key
     */
    public function testGeneric(
        array $callbacks,
        $current,
        $key,
        array $iterator,
        bool $expected
    ): void {
        self::assertSame($expected, CallbacksArrayReducer::or()($callbacks)($current, $key, $iterator));
    }
}
```

</details>

**Comment by maks-rafalko** [comment, marked as answer]
It looks like this line of code is marked as non-covered by underlying coverage driver. Could you please try to generate code coverage HTML report in the same tools (PHP version, PHPUnit and code coverage driver version) and check if this line is covered?

**Comment by drupol** [comment]
Hi !

Thanks for your answer.

Here's the screenshot of the HTML Code Coverage on that file:

![image](https://user-images.githubusercontent.com/252042/167259646-c1ff54c4-be0d-4b99-a852-33dc1c21b049.png)

The line 45 containing `false` is not highlighted in green, do you think that could be the root cause of the issue ?

**Reply by maks-rafalko** [reply]
I'm sure for 100%. This is how it looks on my setup (PHP 8.1, pcov or xdebug):

![image](https://user-images.githubusercontent.com/3725595/167263076-b2837a04-00ae-4d06-8f85-636b467918a0.png)


and I have 100% MSI (all mutants are killed)

**Reply by drupol** [reply]
That's weird.

What are the software version you use?
Are you using pcov or xdebug?
What's your phpunit configuration?

**Reply by maks-rafalko** [reply]
PHP 8.1.4 (cli) (built: Mar 16 2022 11:32:47) (NTS)

pcov

PHPUnit 9.5.7 by Sebastian Bergmann and contributors.

![image](https://user-images.githubusercontent.com/3725595/167264811-16a316d3-8587-42c5-8a47-85405b73bf41.png)

**Reply by drupol** [reply]
Thanks! Going to disable XDebug and test with pcov only.

**Reply by drupol** [reply]
I succeeded to reproduce the use case on the playground: https://infection-php.dev/r/n44w

However, when I run it locally and on the CI, it says: 

```
  - The file "index.xml" could not be found: Could not find any "index.xml" f  
  ile in "/tmp/infection"   
```

Find the log here: https://github.com/loophp/collection/runs/6335797750?check_suite_focus=true
And the relevant PR: https://github.com/loophp/collection/pull/252

**Reply by drupol** [reply]
I think I found the issue !

**Reply by maks-rafalko** [reply]
so, what's the issue?

**Reply by drupol** [reply]
I think that if I add this bit of configuration, `infection` is not able to find `index.xml` anymore, and I couldn't find a way to configure that yet.

```xml
        <report>
            <xml outputDirectory="build/logs/coverage/coverage-xml"/>
            <html outputDirectory="build/logs/coverage/html-coverage"/>
            <clover outputFile="build/logs/coverage/clover.xml"/>
        </report>
```

**Reply by maks-rafalko** [reply]
sounds like a bug in infection itself... if so - please open an issue

**Reply by drupol** [reply]
Apparently, phpunit and pcov doesn't work with using `pathCoverage="true"`. I removed it and changed a bit the paths. I also added `tmpDir` in `infection.json`, see the final diff: https://github.com/loophp/collection/pull/252/files

**Reply by maks-rafalko** [reply]
yes, path coverage works with `xdebug` only. So if you really need path coverage, the only option is `xdebug`. Otherwise `pcov` is a good choice

**Reply by drupol** [reply]
I'm now using pcov, and I have exactly the same issue, locally and on the ci.

![image](https://user-images.githubusercontent.com/252042/167270973-74b6f81f-ea9b-4d12-8a4c-de91328aac23.png)

Investigations ongoing...  once again.

**Reply by maks-rafalko** [reply]
unfortunately from Infection side we can't do much here without some additional tricks/workarounds, because it's an issue of coverage driver.

Other options are to change the source code:

1. move `false` to the line above like `static fn () ..., false` (yeah, I understand that it can be not pretty)
2. assign `false` to variable above and use variable instead of raw `false`. In this case coverage driver should mark variable declaration line of code as "executed"
  ```php
  $false = false;
  static fn (...$parameters): bool => array_reduce(
    $callbacks,
     /**
    * @param bool $carry
    * @param callable(mixed...): bool $callable
    */
    static fn (bool $carry, callable $callable): bool => $carry || $callable(...$parameters),
    $false
  );
  ```
3. ignore this mutant
